torchtune/training/_distributed.py

```diff
@@ -166,64 +179,95 @@ def load_from_full_model_state_dict(
-            # long-term solution is `swap_tensor`
-            mesh = sharded_meta_param.device_mesh
-            if mesh.ndim > 1:
-                raise NotImplementedError(f"only support 1D FSDP but got {mesh.ndim=}")
-            shard_mesh_dim = 0
-            shard_world_size = mesh.size(shard_mesh_dim)
-            shard_rank = cast(
-                torch.distributed.ProcessGroup, mesh.get_group(shard_mesh_dim)
-            ).rank()
-            chunk = list(torch.chunk(full_tensor, shard_world_size, dim=0))[shard_rank]
-            sharded_param = full_tensor.new_zeros(chunk.size())
-            sharded_param[: chunk.size(0)].copy_(chunk)
-
-            # TODO: change to from_local API (need to add view support for NF4)
-            sharded_tensor = DTensor(
-                local_tensor=sharded_param,
-                spec=DTensorSpec(
-                    mesh=sharded_meta_param.device_mesh,
-                    placements=sharded_meta_param.placements,
-                    tensor_meta=TensorMeta(
-                        shape=sharded_meta_param.size(),
-                        dtype=sharded_meta_param.dtype,
-                        stride=sharded_meta_param.stride(),
+    # PyTorch nightly versions from December 20, 2024, support the following features:
+    # - `set_model_state_dict` with the `cpu_offload` option
+    # - Multiple devices in local state dict
+    # - Relative optimizations for improved memory performance
+    # Please keep the version check `_USE_DISTRIBUTED_STATE_DICT_API` until these changes are
+    # released in the PyTorch stable version.
+    has_nf4 = any(
+        hasattr(param, "_local_tensor") and isinstance(param._local_tensor, NF4Tensor)
+        for param in model.parameters()
+    )
+    if _USE_DISTRIBUTED_STATE_DICT_API and not has_nf4:
+        meta_sharded_sd = model.state_dict()
+        for param_name in full_sd.keys():
+            sharded_meta_param = meta_sharded_sd.get(param_name)
+            full_sd[param_name] = full_sd[param_name].to(sharded_meta_param.dtype)
+        options = StateDictOptions(
+            full_state_dict=True,
+            broadcast_from_rank0=True,
+            strict=strict,
+            cpu_offload=cpu_offload,
+        )
+        return set_model_state_dict(
+            model=model, model_state_dict=full_sd, options=options
+        )
+    else:
+        meta_sharded_sd = model.state_dict()
+        sharded_sd = {}
+        for param_name, full_tensor in full_sd.items():
+            sharded_meta_param = meta_sharded_sd.get(param_name)
+            full_tensor = full_tensor.to(sharded_meta_param.dtype).to(device)
+            if hasattr(sharded_meta_param, "_local_tensor") and isinstance(
+                sharded_meta_param._local_tensor, NF4Tensor
+            ):
+                block_size = sharded_meta_param._local_tensor.block_size
+                scaler_block_size = sharded_meta_param._local_tensor.scaler_block_size
+                full_tensor = to_nf4(
```

Comment:
if we assume "full" means plain tensor which is the claim above, then this is - in fact - not a "full" tensor?